lib/Dialect/FIRRTL/LowerTypes.cpp

```diff
@@ -0,0 +1,485 @@
+  circt::DenseMap<Value, ValueField> loweredSubfieldInfo;
+};
+} // end anonymous namespace
+
+/// This is the pass constructor.
+std::unique_ptr<mlir::Pass> circt::firrtl::createLowerFIRRTLTypesPass() {
+  return std::make_unique<FIRRTLTypesLowering>();
+}
+
+// This is the main entrypoint for the lowering pass.
+void FIRRTLTypesLowering::runOnOperation() {
+  auto module = getOperation();
+  auto *body = module.getBodyBlock();
+
+  OpBuilder theBuilder(&getContext());
+  builder = &theBuilder;
+
+  // Lower the module block arguments.
+  SmallVector<BlockArgument, 8> args(body->args_begin(), body->args_end());
+  for (auto arg : args)
+    mlir::TypeSwitch<Type>(arg.getType())
+        .Case<BundleType>([&](BundleType type) { lowerArg(arg, type); })
+        .Default<>([&](Type type) { lowerArg(arg, type); });
+
+  // Lower the operations.
+  for (auto &op : body->getOperations()) {
+    builder->setInsertionPoint(&op);
+    if (failed(dispatchVisitor(&op)))
+      return;
+  }
+
+  cleanup();
+}
+
+//===----------------------------------------------------------------------===//
+// Lowering module block arguments.
+//===----------------------------------------------------------------------===//
+
+// Lower arguments with bundle type by flattening them.
+void FIRRTLTypesLowering::lowerArg(BlockArgument arg, BundleType type) {
+  // Flatten the bundle types.
+  SmallVector<FlatBundleFieldEntry, 8> fieldTypes;
+  flattenBundleTypes(type, "", false, fieldTypes);
+
+  for (auto field : fieldTypes) {
+    // Create flattened block arguments.
+    auto type = getPortType(field);
+    auto newValue = addArgument(type, arg.getArgNumber(), field.suffix);
+
+    // Map the flattened suffix for the original argument to the new value.
+    setBundleLowering(arg, field.suffix, newValue);
+  }
+}
+
+// Lower any other arguments by copying them to keep the relative order.
+void FIRRTLTypesLowering::lowerArg(BlockArgument arg, Type type) {
+  auto newValue = addArgument(type, arg.getArgNumber());
+  arg.replaceAllUsesWith(newValue);
+}
+
+//===----------------------------------------------------------------------===//
+// Lowering operations.
+//===----------------------------------------------------------------------===//
+
+// Lower instance operations in the same way as module block arguments. Bundles
+// are flattened, and other arguments are copied to keep the relative order. By
+// ensuring both lowerings are the same, we can process every module in the
+// circuit in parallel, and every instance will have the correct ports.
+LogicalResult FIRRTLTypesLowering::visitDecl(InstanceOp op) {
+  // The instance's ports are represented as one value with a bundle type.
+  BundleType originalType = op.result().getType().cast<BundleType>();
+
+  // Create a new, flat bundle type for the new result
+  SmallVector<BundleType::BundleElement, 8> bundleElements;
+  for (auto element : originalType.getElements()) {
+    if (element.second.isa<BundleType>()) {
```

Comment:
Same comment, I think you can remove the "if bundle" from the body of the loop since flattenBundleTypes is correct for non-bundles.